modules/porous_flow/src/kernels/PorousFlowAdvectiveFlux.C

```diff
@@ -0,0 +1,321 @@
+  _fluid_viscosity(getMaterialProperty<std::vector<Real> >("PorousFlow_viscosity")),
+  _dfluid_viscosity_dvar(getMaterialProperty<std::vector<std::vector<Real> > >("dPorousFlow_viscosity_dvar")),
+  _mass_fractions(getMaterialProperty<std::vector<std::vector<Real> > >("PorousFlow_mass_frac")),
+  _dmass_fractions_dvar(getMaterialProperty<std::vector<std::vector<std::vector<Real> > > >("dPorousFlow_mass_frac_dvar")),
+  _pp(getMaterialProperty<std::vector<Real> >("PorousFlow_porepressure_nodal")),
+  _grad_p(getMaterialProperty<std::vector<RealGradient> >("PorousFlow_grad_porepressure_qp")),
+  _dgrad_p_dgrad_var(getMaterialProperty<std::vector<std::vector<Real> > >("dPorousFlow_grad_porepressure_qp_dgradvar")),
+  _dgrad_p_dvar(getMaterialProperty<std::vector<std::vector<RealGradient> > >("dPorousFlow_grad_porepressure_qp_dvar")),
+  _relative_permeability(getMaterialProperty<std::vector<Real> >("PorousFlow_relative_permeability")),
+  _drelative_permeability_dvar(getMaterialProperty<std::vector<std::vector<Real> > >("dPorousFlow_relative_permeability_dvar")),
+  _porousflow_dictator_UO(getUserObject<PorousFlowDictator>("PorousFlowDictator_UO")),
+  _component_index(getParam<unsigned int>("component_index")),
+  _gravity(getParam<RealVectorValue>("gravity"))
+{
+  // Make sure that this kernels variable is a valid PorousFlow variable
+  if (_porousflow_dictator_UO.not_porflow_var(_var.number()))
+    mooseError(" Variable " << _var.name() << " in the " << _name << " kernel is not a valid PorousFlow variable");
+
+  _num_phases = _porousflow_dictator_UO.num_phases();
+}
+
+Real
+PorousFlowAdvectiveFlux::darcyQp(unsigned int ph)
+{
+  return _grad_test[_i][_qp]*(_permeability[_qp]*(_grad_p[_qp][ph] - _fluid_density_qp[_qp][ph]*_gravity));
+}
+
+Real PorousFlowAdvectiveFlux::darcyQpJacobian(unsigned int jvar, unsigned int ph)
+{
+  if (_porousflow_dictator_UO.not_porflow_var(jvar))
+    return 0.0;
+
+  const unsigned int pvar = _porousflow_dictator_UO.porflow_var_num(jvar);
+  return _grad_test[_i][_qp] * (_dpermeability_dvar[_qp][pvar] * (_grad_p[_qp][ph] - _fluid_density_qp[_qp][ph]*_gravity) + _permeability[_qp] * (_grad_phi[_j][_qp] * _dgrad_p_dgrad_var[_qp][ph][pvar] - _phi[_j][_qp] * _dfluid_density_qp_dvar[_qp][ph][pvar] * _gravity) + _permeability[_qp]*(_dgrad_p_dvar[_qp][ph][pvar]*_phi[_j][_qp]) );
+}
+
+
+Real PorousFlowAdvectiveFlux::computeQpResidual()
+{
+  mooseError("PorousFlowAdvectiveFlux: computeQpResidual called");
+  return 0.0;
+}
+
+
+void PorousFlowAdvectiveFlux::computeResidual()
+{
+  upwind(true, false, 0.);
+}
+
+void PorousFlowAdvectiveFlux::computeJacobian()
+{
+   upwind(false, true, _var.number());
+}
+
+void PorousFlowAdvectiveFlux::computeOffDiagJacobian(unsigned int jvar)
+{
+   upwind(false, true, jvar);
+}
+
+
+void PorousFlowAdvectiveFlux::upwind(bool compute_res, bool compute_jac, unsigned int jvar)
```

Comment:
I'd make this take an enum that is defined like

`enum JacRes { RESIDUAL=1, JACOBIAN=2 }` 

then your calls could be `upwind(RESIDUAL, jvar);`,  `upwind(JACOBIAN, jvar);` , or  `upwind(RESIDUAL + JACOBIAN, jvar);` . No big deal, as this is only an internal interface, but the meaning of the bool params is obscure when you read the calls to upwind.